src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -156,9 +150,73 @@ async Task AddDocumentsFromProjectAsync(Project? project, ImmutableArray<string>
+                if (isFSAOn)
+                {
+                    result.Add(new ProjectDiagnosticSource(project));
                 }
             }
         }
+
+        private record struct ProjectDiagnosticSource(Project Project) : IDiagnosticSource
+        {
+            public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+                IDiagnosticAnalyzerService diagnosticAnalyzerService,
+                RequestContext context,
+                DiagnosticMode diagnosticMode,
+                CancellationToken cancellationToken)
+            {
+                // Directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
+                // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
+                // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
+                // and do not need to be adjusted.
+                var projectDiagnostics = await diagnosticAnalyzerService.GetProjectDiagnosticsForIdsAsync(Project.Solution, Project.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
+                return projectDiagnostics;
+            }
+
+            public ProjectOrDocumentId GetId() => new(Project.Id);
+
+            public Project GetProject() => Project;
+
+            public Uri GetUri()
+            {
+                Contract.ThrowIfNull(Project.FilePath);
+                return ProtocolConversions.GetUriFromFilePath(Project.FilePath);
+            }
+        }
+
+        private record struct WorkspaceDocumentDiagnosticSource(Document Document) : IDiagnosticSource
+        {
+            public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+                IDiagnosticAnalyzerService diagnosticAnalyzerService,
+                RequestContext context,
+                DiagnosticMode diagnosticMode,
+                CancellationToken cancellationToken)
+            {
+                if (Document is not SourceGeneratedDocument)
+                {
+                    // We call GetDiagnosticsForIdsAsync as we want to ensure we get the full set of diagnostics for this document
+                    // including those reported as a compilation end diagnostic.  These are not included in document pull (uses GetDiagnosticsForSpan) due to cost.
+                    // However we can include them as a part of workspace pull when FSA is on.
+                    var documentDiagnostics = await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(Document.Project.Solution, Document.Project.Id, Document.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    return documentDiagnostics;
+                }
+                else
+                {
+                    // Unfortunately GetDiagnosticsForIdsAsync returns nothing for source generated documents.
```

Comment:
@jasonmalinowski @mavasani is this intentional?  Seems weird since technically GetDiagnosticsForIdsAsync should be the slow version.